### PR TITLE
Vertex state validation

### PR DESCRIPTION
**Pull request by Kangz**
PTAL! Each commit should be easy to review individually (except maybe the last one, but it just contains a bunch of new tests).

-----

- [X] New helpers, if any, are documented in `helper_index.md`.
- [X] Incomplete tests, if any, are marked with TODO or `.unimplemented()`.

**[Review requirement](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) checklist:**

- [x] WebGPU readability
- [x] TypeScript readability

### CODE REVIEW

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 1]
```diff
@@ -488,183 +459,202 @@ g.test('vertex_shader_input_location_in_vertex_state')
...
+        const offsetsToTest = [0, bytesPerComponent, p.arrayStride - formatSize];
+
+        // Avoid adding duplicate offsets because the harness will assert each case is different.
+        const halfAlignment = Math.floor(bytesPerComponent / 2);
+        if (halfAlignment !== 0) {
+          offsetsToTest.push(halfAlignment);
+          offsetsToTest.push(p.arrayStride - formatSize - halfAlignment);
+        }
+
+        return poptions('offset', offsetsToTest);
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:540]: nit: Could do something like
```ts
        return poptions(
          'offset',
          new Set([
            0,
            halfAlignment,
            bytesPerComponent,
            p.arrayStride - formatSize,
            p.arrayStride - formatSize - halfAlignment,
          ])
        );
```
I think.

OTOH, you could filter out formats with bytesPerComponent of 1. Aren't they kind of pointless for this test?

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 2]
```diff
@@ -488,183 +459,202 @@ g.test('vertex_shader_input_location_in_vertex_state')
...
+    const formatInfo = kVertexFormatInfo[format];
+    const formatSize = formatInfo.bytesPerComponent * formatInfo.componentCount;
+    const limit = arrayStride === 0 ? kMaxVertexBufferArrayStride : arrayStride;
+
+    const success = offset + formatSize <= limit;
+    t.testVertexState(success, { vertexBuffers });
+  });
+
+g.test('many_attributes_overlapping')
+  .desc(`Test that it is valid to have many vertex attributes overlap`)
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:649]: nit: Maybe test with different formats (like rotate the `format` between the attributes)?

Or maybe this should be an operation test later that actually tests the load behavior? (or both validation and operation test)

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 3]
```diff
@@ -488,183 +459,202 @@ g.test('vertex_shader_input_location_in_vertex_state')
...
+    Test that vertex attribute [offset, offset + formatSize) must be contained in the arrayStride if arrayStride is not 0:
+    - Test for all formats.
+    - Test for various arrayStrides and offsets within that stride
+    - Test for various vertex buffer indices
+    - Test for various amounts of attributes in that vertex buffer`
+  )
+  .cases(
+    params()
+      .combine(poptions('format', kVertexFormats))
+      .combine(
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:586]: nit: Probably arrayStride and offset (at least offset) make more sense as subcases

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 4]
```diff
@@ -488,183 +459,202 @@ g.test('vertex_shader_input_location_in_vertex_state')
...
+        ])
+      )
+      .expand(p => {
+        // Compute a bunch of test offsets to test.
+        const { bytesPerComponent, componentCount } = kVertexFormatInfo[p.format];
+        const formatSize = bytesPerComponent * componentCount;
+        const offsetsToTest = [0, bytesPerComponent];
+
+        // Avoid adding offset based on the arrayStride if it would make duplicated tests at it
+        // makes the harness assert.
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:601]: I don't think this comment is applicable here?

**Review by kainino0x — APPROVED** [review]

File: src/webgpu/api/validation/vertex_state.spec.ts (hunk 1 above)
Comment (Kangz) [reply src/webgpu/api/validation/vertex_state.spec.ts:540]: Went with the Set approach.

File: src/webgpu/api/validation/vertex_state.spec.ts (hunk 2 above)
Comment (Kangz) [reply src/webgpu/api/validation/vertex_state.spec.ts:649]: There should be an operation test that does complicated cases like this, but I still added the format rotation.

File: src/webgpu/api/validation/vertex_state.spec.ts (hunk 3 above)
Comment (kainino0x) [reply src/webgpu/api/validation/vertex_state.spec.ts:587]: It's a little weird but it can be done, see the commit I pushed

Comment (kainino0x) [reply src/webgpu/api/validation/vertex_state.spec.ts:587]: (I have ideas about how to improve this - subcases could be passed a pre-constructed single-value params() object so you can just `p => p.combine`, but haven't had a chance to try this)

**Review by kainino0x — APPROVED** [review]
nit: prefer to push new commits instead of force-pushing, to make re-review easier (though it was ok in this case because there was no rebase / the "force-push" diff on github was clean)

**Comment by Kangz** [comment]
Yeah I should avoid force push. What do you think of my adding small commits like "fix for X, fix for Foo" then I'll do a rebase -i before and force-push before merging?

File: src/webgpu/api/validation/vertex_state.spec.ts (hunk 3 above)
Comment (Kangz) [reply src/webgpu/api/validation/vertex_state.spec.ts:587]: Thanks for the fix!

**Comment by kainino0x** [comment]
I use squash-and-merge for almost all PRs, but if you want multiple commits in a PR that sounds good, just let me know so I won't land it.

**Pull request merged**